### ISSUE TITLE
Return `Result<OpaqueRust, OpaqueRust>` from an async function

### ISSUE THREAD

**Issue by chinedufn** [chinedufn/swift-bridge#157]
Right now the following signature leads to a compile time error:

```rust
#[swift_bridge::bridge]
mod ffi {
    extern "Rust" {
        type TypeA;
        type TypeB;

        async fn some_function() -> Result<TypeA, TypeB>;
    }
}
```

We should implement support for returning a `Result<OpaqueRust, OpaqueRust>` from an async Rust function.

### Returning a result

In #149 we added the ability to return an opaque Rust type.

### Integration tests

Here's an example of an integration test for returning a Result from a non async function.

https://github.com/chinedufn/swift-bridge/blob/2ab37cbb949be4a466848129e7f17ecbb3fecf5e/crates/swift-integration-tests/src/result.rs#L6-L8

https://github.com/chinedufn/swift-bridge/blob/2ab37cbb949be4a466848129e7f17ecbb3fecf5e/SwiftRustIntegrationTestRunner/SwiftRustIntegrationTestRunnerTests/ResultTests.swift#L17-L32

---

And here's an example of an async integration test:

https://github.com/chinedufn/swift-bridge/blob/dc1119e1ee333e1412f81f726987637b8f620a58/crates/swift-integration-tests/src/async_function.rs#L8

https://github.com/chinedufn/swift-bridge/blob/dc1119e1ee333e1412f81f726987637b8f620a58/SwiftRustIntegrationTestRunner/SwiftRustIntegrationTestRunnerTests/AsyncTests.swift#L29-L33

### Result codegen test

Here's an example of a codegen for a non async function that returns a Result:

https://github.com/chinedufn/swift-bridge/blob/2ab37cbb949be4a466848129e7f17ecbb3fecf5e/crates/swift-bridge-ir/src/codegen/codegen_tests/result_codegen_tests.rs#L129-L194

### Async codegen test

Here's an example of an async codegen test where we return a value:

https://github.com/chinedufn/swift-bridge/blob/dc1119e1ee333e1412f81f726987637b8f620a58/crates/swift-bridge-ir/src/codegen/codegen_tests/async_function_codegen_tests.rs#L270-L355

**Comment by NiwakaDev** [comment]
I want to try to implement this!

**Comment by chinedufn** [comment]
Go for it!

**Comment by NiwakaDev** [comment]
Hi, @chinedufn. I have a quetion. 

As far as I understand, Do we manually implement Swift's `Error` that `E` conforms to?  

※`E` is `Result<T, E>`'s variant in Swift.

**Comment by chinedufn** [comment]
Yeah for now we need to do it manually until we address https://github.com/chinedufn/swift-bridge/issues/150

**Comment by chinedufn** [comment]
Closed via #158
